File: web/src/world.ts
export type WorldBounds = {
  min_x: number;
  max_x: number;
  min_y: number;
  max_y: number;
};

export type RobotSpawn = {
  x_m: number;
  y_m: number;
  yaw_deg: number;
};

export type WorldObstacle = {
  id: string;
  center_x_m: number;
  center_y_m: number;
  size_x_m: number;
  size_y_m: number;
  height_m: number;
};

export type SemanticObjectKind = "person" | "table" | "chair" | "bottle" | "door" | "goal_zone";
export type SemanticObjectColor = "red" | "blue" | "green" | "yellow" | "gray" | "white";

export type WorldSemanticObject = {
  id: string;
  kind: SemanticObjectKind;
  label: string;
  color: SemanticObjectColor;
  center_x_m: number;
  center_y_m: number;
  size_x_m: number;
  size_y_m: number;
  height_m: number;
  interaction_radius_m: number;
  blocking: boolean;
};

export type WorldConfig = {
  version: 1;
  scene_id: string;
  bounds: WorldBounds;
  robot_spawn: RobotSpawn;
  robot_footprint_radius_m: number;
  stop_clearance_m: number;
  sensor_max_range_m: number;
  obstacles: WorldObstacle[];
  semantic_objects: WorldSemanticObject[];
};

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

function hasExactKeys(value: Record<string, unknown>, keys: readonly string[]): boolean {
  const expected = new Set(keys);
  return Object.keys(value).length === keys.length && Object.keys(value).every((key) => expected.has(key));
}

function isFiniteNumber(value: unknown): value is number {
  return typeof value === "number" && Number.isFinite(value);
}

function isIdentifier(value: unknown): value is string {
  return typeof value === "string" && /^[A-Za-z0-9][A-Za-z0-9_.:-]{0,63}$/.test(value);
}

function parseBounds(value: unknown): WorldBounds | null {
  if (!isRecord(value) || !hasExactKeys(value, ["min_x", "max_x", "min_y", "max_y"])) return null;
  const minX = value.min_x;
  const maxX = value.max_x;
  const minY = value.min_y;
  const maxY = value.max_y;
  if (!isFiniteNumber(minX) || !isFiniteNumber(maxX) || !isFiniteNumber(minY) || !isFiniteNumber(maxY)) {
    return null;
  }
  if (minX >= maxX || minY >= maxY) return null;
  return {
    min_x: minX,
    max_x: maxX,
    min_y: minY,
    max_y: maxY,
  };
}

function parseSpawn(value: unknown): RobotSpawn | null {
  if (!isRecord(value) || !hasExactKeys(value, ["x_m", "y_m", "yaw_deg"])) return null;
  const x = value.x_m;
  const y = value.y_m;
  const yaw = value.yaw_deg;
  if (!isFiniteNumber(x) || !isFiniteNumber(y) || !isFiniteNumber(yaw)) return null;
  return { x_m: x, y_m: y, yaw_deg: yaw };
}

function parseObstacle(value: unknown): WorldObstacle | null {
  const keys = ["id", "center_x_m", "center_y_m", "size_x_m", "size_y_m", "height_m"] as const;
  if (!isRecord(value) || !hasExactKeys(value, keys)) return null;
  if (!isIdentifier(value.id)) return null;
  const centerX = value.center_x_m;
  const centerY = value.center_y_m;
  const sizeX = value.size_x_m;
  const sizeY = value.size_y_m;
  const height = value.height_m;
  if (
    !isFiniteNumber(centerX)
    || !isFiniteNumber(centerY)
    || !isFiniteNumber(sizeX)
    || !isFiniteNumber(sizeY)
    || !isFiniteNumber(height)
  ) {
    return null;
  }
  if (sizeX <= 0 || sizeY <= 0 || height <= 0) return null;
  return {
    id: value.id,
    center_x_m: centerX,
    center_y_m: centerY,
    size_x_m: sizeX,
    size_y_m: sizeY,
    height_m: height,
  };
}

const semanticKinds = new Set<SemanticObjectKind>([
  "person",
  "table",
  "chair",
  "bottle",
  "door",
  "goal_zone",
]);
const semanticColors = new Set<SemanticObjectColor>(["red", "blue", "green", "yellow", "gray", "white"]);

function parseSemanticObject(value: unknown): WorldSemanticObject | null {
  const keys = [
    "id",
    "kind",
    "label",
    "color",
    "center_x_m",
    "center_y_m",
    "size_x_m",
    "size_y_m",
    "height_m",
    "interaction_radius_m",
    "blocking",
  ] as const;
  if (!isRecord(value) || !hasExactKeys(value, keys) || !isIdentifier(value.id)) return null;
  if (typeof value.kind !== "string" || !semanticKinds.has(value.kind as SemanticObjectKind)) return null;
  if (typeof value.color !== "string" || !semanticColors.has(value.color as SemanticObjectColor)) return null;
  if (typeof value.label !== "string" || value.label.trim().length === 0 || value.label.length > 64) return null;
  if (typeof value.blocking !== "boolean") return null;
  const centerX = value.center_x_m;
  const centerY = value.center_y_m;
  const sizeX = value.size_x_m;
  const sizeY = value.size_y_m;
  const height = value.height_m;
  const interactionRadius = value.interaction_radius_m;
  if (
    !isFiniteNumber(centerX)
    || !isFiniteNumber(centerY)
    || !isFiniteNumber(sizeX)
    || !isFiniteNumber(sizeY)
    || !isFiniteNumber(height)
    || !isFiniteNumber(interactionRadius)
  ) {
    return null;
  }
  if (sizeX <= 0 || sizeY <= 0 || height <= 0 || interactionRadius <= 0 || interactionRadius > 2) return null;
  return {
    id: value.id,
    kind: value.kind as SemanticObjectKind,
    label: value.label.trim(),
    color: value.color as SemanticObjectColor,
    center_x_m: centerX,
    center_y_m: centerY,
    size_x_m: sizeX,
    size_y_m: sizeY,
    height_m: height,
    interaction_radius_m: interactionRadius,
    blocking: value.blocking,
  };
}

function pointToRectDistance(x: number, y: number, obstacle: WorldObstacle): number {
  const halfX = obstacle.size_x_m / 2;
  const halfY = obstacle.size_y_m / 2;
  const dx = Math.max(Math.abs(x - obstacle.center_x_m) - halfX, 0);
  const dy = Math.max(Math.abs(y - obstacle.center_y_m) - halfY, 0);
  return Math.hypot(dx, dy);
}

export function parseWorldConfig(value: unknown): WorldConfig | null {
  const keys = [
    "version",
    "scene_id",
    "bounds",
    "robot_spawn",
    "robot_footprint_radius_m",
    "stop_clearance_m",
    "sensor_max_range_m",
    "obstacles",
    "semantic_objects",
  ] as const;
  if (!isRecord(value) || !hasExactKeys(value, keys) || value.version !== 1 || !isIdentifier(value.scene_id)) {
    return null;
  }
  const bounds = parseBounds(value.bounds);
  const robotSpawn = parseSpawn(value.robot_spawn);
  if (!bounds || !robotSpawn) return null;
  if (!isFiniteNumber(value.robot_footprint_radius_m) || value.robot_footprint_radius_m <= 0) return null;
  if (!isFiniteNumber(value.stop_clearance_m) || value.stop_clearance_m < 0) return null;
  if (!isFiniteNumber(value.sensor_max_range_m) || value.sensor_max_range_m <= 0) return null;
  if (!Array.isArray(value.obstacles) || !Array.isArray(value.semantic_objects)) return null;

  const obstacles: WorldObstacle[] = [];
  const ids = new Set<string>();
  for (const rawObstacle of value.obstacles) {
    const obstacle = parseObstacle(rawObstacle);
    if (!obstacle || ids.has(obstacle.id)) return null;
    ids.add(obstacle.id);
    if (
      obstacle.center_x_m - obstacle.size_x_m / 2 < bounds.min_x
      || obstacle.center_x_m + obstacle.size_x_m / 2 > bounds.max_x
      || obstacle.center_y_m - obstacle.size_y_m / 2 < bounds.min_y
      || obstacle.center_y_m + obstacle.size_y_m / 2 > bounds.max_y
    ) {
      return null;
    }
    obstacles.push(obstacle);
  }

  const semanticObjects: WorldSemanticObject[] = [];
  for (const rawItem of value.semantic_objects) {
    const item = parseSemanticObject(rawItem);
    if (!item || ids.has(item.id)) return null;
    ids.add(item.id);
    if (
      item.center_x_m - item.size_x_m / 2 < bounds.min_x
      || item.center_x_m + item.size_x_m / 2 > bounds.max_x
      || item.center_y_m - item.size_y_m / 2 < bounds.min_y
      || item.center_y_m + item.size_y_m / 2 > bounds.max_y
    ) {
      return null;
    }
    semanticObjects.push(item);
  }

  const margin = value.robot_footprint_radius_m + value.stop_clearance_m;
  if (
    bounds.max_x - bounds.min_x <= margin * 2
    || bounds.max_y - bounds.min_y <= margin * 2
    || robotSpawn.x_m - margin < bounds.min_x
    || robotSpawn.x_m + margin > bounds.max_x
    || robotSpawn.y_m - margin < bounds.min_y
    || robotSpawn.y_m + margin > bounds.max_y
    || obstacles.some((obstacle) => pointToRectDistance(robotSpawn.x_m, robotSpawn.y_m, obstacle) <= margin)
    || semanticObjects.some(
      (item) => item.blocking && pointToRectDistance(robotSpawn.x_m, robotSpawn.y_m, item) <= margin,
    )
  ) {
    return null;
  }
  return {
    version: 1,
    scene_id: value.scene_id,
    bounds,
    robot_spawn: robotSpawn,
    robot_footprint_radius_m: value.robot_footprint_radius_m,
    stop_clearance_m: value.stop_clearance_m,
    sensor_max_range_m: value.sensor_max_range_m,
    obstacles,
    semantic_objects: semanticObjects,
  };
}

File: src/embodied_agent/simulation/world.py
"""Versioned world configuration and deterministic 2D spatial geometry."""

from __future__ import annotations

import math
from typing import Literal, TypeAlias

from pydantic import BaseModel, ConfigDict, Field, StrictBool, StrictStr, field_validator, model_validator
from shapely.geometry import LineString, Point, box
from shapely.geometry.base import BaseGeometry
from shapely.ops import nearest_points

from .models import SimulationSensors


class WorldBounds(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True, allow_inf_nan=False)

    min_x: float
    max_x: float
    min_y: float
    max_y: float

    @model_validator(mode="after")
    def validate_order(self) -> "WorldBounds":
        if self.min_x >= self.max_x or self.min_y >= self.max_y:
            raise ValueError("world bounds minima must be smaller than maxima")
        return self


class RobotSpawn(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True, allow_inf_nan=False)

    x_m: float = 0.0
    y_m: float = 0.0
    yaw_deg: float = 0.0


class WorldObstacle(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True, allow_inf_nan=False)

    id: str = Field(..., min_length=1, max_length=64, pattern=r"^[A-Za-z0-9][A-Za-z0-9_.:-]{0,63}$")
    center_x_m: float
    center_y_m: float
    size_x_m: float = Field(..., gt=0.0)
    size_y_m: float = Field(..., gt=0.0)
    height_m: float = Field(default=0.7, gt=0.0)

    def polygon(self) -> BaseGeometry:
        half_x = self.size_x_m / 2.0
        half_y = self.size_y_m / 2.0
        return box(
            self.center_x_m - half_x,
            self.center_y_m - half_y,
            self.center_x_m + half_x,
            self.center_y_m + half_y,
        )


SemanticObjectKind: TypeAlias = Literal[
    "person",
    "table",
    "chair",
    "bottle",
    "door",
    "goal_zone",
]
SemanticObjectColor: TypeAlias = Literal["red", "blue", "green", "yellow", "gray", "white"]


class WorldSemanticObject(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True, allow_inf_nan=False)

    id: str = Field(..., min_length=1, max_length=64, pattern=r"^[A-Za-z0-9][A-Za-z0-9_.:-]{0,63}$")
    kind: SemanticObjectKind
    label: StrictStr = Field(..., min_length=1, max_length=64)
    color: SemanticObjectColor
    center_x_m: float
    center_y_m: float
    size_x_m: float = Field(..., gt=0.0)
    size_y_m: float = Field(..., gt=0.0)
    height_m: float = Field(..., gt=0.0)
    interaction_radius_m: float = Field(..., gt=0.0, le=2.0)
    blocking: StrictBool

    @field_validator("label")
    @classmethod
    def normalize_label(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized:
            raise ValueError("semantic object label must not be blank")
        return normalized

    def polygon(self) -> BaseGeometry:
        half_x = self.size_x_m / 2.0
        half_y = self.size_y_m / 2.0
        return box(
            self.center_x_m - half_x,
            self.center_y_m - half_y,
            self.center_x_m + half_x,
            self.center_y_m + half_y,
        )


class WorldConfig(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True, allow_inf_nan=False)

    version: Literal[1] = 1
    scene_id: str = Field(..., min_length=1, max_length=64, pattern=r"^[A-Za-z0-9][A-Za-z0-9_.:-]{0,63}$")
    bounds: WorldBounds
    robot_spawn: RobotSpawn = RobotSpawn()
    robot_footprint_radius_m: float = Field(default=0.35, gt=0.0)
    stop_clearance_m: float = Field(default=0.25, ge=0.0)
    sensor_max_range_m: float = Field(default=10.0, gt=0.0)
    obstacles: tuple[WorldObstacle, ...] = ()
    semantic_objects: tuple[WorldSemanticObject, ...] = ()

    @model_validator(mode="after")
    def validate_geometry(self) -> "WorldConfig":
        margin = self.robot_footprint_radius_m + self.stop_clearance_m
        if self.bounds.max_x - self.bounds.min_x <= 2.0 * margin:
            raise ValueError("world x bounds are too small for the robot footprint")
        if self.bounds.max_y - self.bounds.min_y <= 2.0 * margin:
            raise ValueError("world y bounds are too small for the robot footprint")

        ids = [obstacle.id for obstacle in self.obstacles]
        ids.extend(item.id for item in self.semantic_objects)
        if len(ids) != len(set(ids)):
            raise ValueError("world object ids must be globally unique")

        world_polygon = box(
            self.bounds.min_x,
            self.bounds.min_y,
            self.bounds.max_x,
            self.bounds.max_y,
        )
        safe_spawn = Point(self.robot_spawn.x_m, self.robot_spawn.y_m).buffer(margin)
        if not world_polygon.covers(safe_spawn):
            raise ValueError("robot spawn and clearance must stay inside world bounds")

        for obstacle in self.obstacles:
            polygon = obstacle.polygon()
            if not world_polygon.covers(polygon):
                raise ValueError(f"world obstacle {obstacle.id!r} must stay inside bounds")
            if safe_spawn.intersects(polygon):
                raise ValueError(f"robot spawn conflicts with obstacle {obstacle.id!r}")
        for item in self.semantic_objects:
            polygon = item.polygon()
            if not world_polygon.covers(polygon):
                raise ValueError(f"semantic object {item.id!r} must stay inside bounds")
            if item.blocking and safe_spawn.intersects(polygon):
                raise ValueError(f"robot spawn conflicts with semantic object {item.id!r}")
        return self


class WorldGeometry:
    """Raycast sensors and constrain straight motion against one World Config."""

    _EPSILON_M = 1e-7

    def __init__(self, config: WorldConfig) -> None:
        self.config = config
        radius = config.robot_footprint_radius_m
        motion_margin = radius + config.stop_clearance_m
        self._sensor_bounds = box(
            config.bounds.min_x + radius,
            config.bounds.min_y + radius,
            config.bounds.max_x - radius,
            config.bounds.max_y - radius,
        )
        self._motion_bounds = box(
            config.bounds.min_x + motion_margin,
            config.bounds.min_y + motion_margin,
            config.bounds.max_x - motion_margin,
            config.bounds.max_y - motion_margin,
        )
        solid_objects = (
            *config.obstacles,
            *(item for item in config.semantic_objects if item.blocking),
        )
        self._sensor_obstacles = tuple(
            (obstacle, obstacle.polygon().buffer(radius)) for obstacle in solid_objects
        )
        self._motion_obstacles = tuple(
            (obstacle, obstacle.polygon().buffer(motion_margin))
            for obstacle in solid_objects
        )

    def sensor_distances(self, x_m: float, y_m: float, yaw_deg: float) -> SimulationSensors:
        return SimulationSensors(
            front_distance_cm=self._clean_cm(self._ray_distance_m(x_m, y_m, yaw_deg)),
            left_distance_cm=self._clean_cm(self._ray_distance_m(x_m, y_m, yaw_deg + 90.0)),
            right_distance_cm=self._clean_cm(self._ray_distance_m(x_m, y_m, yaw_deg - 90.0)),
        )

    def constrain_move(
        self,
        x_m: float,
        y_m: float,
        yaw_deg: float,
        distance_m: float,
    ) -> tuple[float, Literal["front_obstacle", "path_obstacle", "world_boundary"] | None]:
        if distance_m == 0.0:
            return 0.0, None

        direction_deg = yaw_deg if distance_m > 0.0 else yaw_deg + 180.0
        desired = abs(distance_m)
        direction = self._direction(direction_deg)
        start = Point(x_m, y_m)
        probe_distance = min(desired, self._EPSILON_M)
        probe = Point(x_m + direction[0] * probe_distance, y_m + direction[1] * probe_distance)
        target = Point(x_m + direction[0] * desired, y_m + direction[1] * desired)
        path = LineString([(probe.x, probe.y), (target.x, target.y)])

        allowed = desired
        reason: Literal["front_obstacle", "path_obstacle", "world_boundary"] | None = None

        if not self._motion_bounds.covers(probe):
            allowed = 0.0
            reason = "world_boundary"
        elif not self._motion_bounds.covers(target):
            boundary_hit = self._intersection_distance(start, path.intersection(self._motion_bounds.boundary))
            if boundary_hit is not None:
                allowed = boundary_hit
                reason = "world_boundary"

        obstacle_reason: Literal["front_obstacle", "path_obstacle"] = (
            "front_obstacle" if distance_m > 0.0 else "path_obstacle"
        )
        for _, forbidden in self._motion_obstacles:
            collision_area = forbidden.buffer(-self._EPSILON_M)
            if collision_area.covers(probe):
                hit_distance = 0.0
            else:
                hit_distance = self._intersection_distance(start, path.intersection(collision_area))
                if hit_distance is not None:
                    hit_distance = max(0.0, hit_distance - self._EPSILON_M)
            if hit_distance is not None and hit_distance < allowed:
                allowed = hit_distance
                reason = obstacle_reason

        if desired <= allowed + self._EPSILON_M:
            return distance_m, None
        signed_allowed = math.copysign(max(0.0, allowed), distance_m)
        return self._clean_m(signed_allowed), reason

    def _ray_distance_m(self, x_m: float, y_m: float, yaw_deg: float) -> float:
        direction = self._direction(yaw_deg)
        max_range = self.config.sensor_max_range_m
        start = Point(x_m, y_m)
        end = Point(x_m + direction[0] * max_range, y_m + direction[1] * max_range)
        ray = LineString([(start.x, start.y), (end.x, end.y)])
        nearest = max_range

        boundary_distance = self._intersection_distance(start, ray.intersection(self._sensor_bounds.boundary))
        if boundary_distance is not None:
            nearest = min(nearest, boundary_distance)

        for _, expanded in self._sensor_obstacles:
            distance = self._intersection_distance(start, ray.intersection(expanded))
            if distance is not None:
                nearest = min(nearest, distance)
        return max(0.0, nearest)

    @staticmethod
    def _intersection_distance(start: Point, geometry: BaseGeometry) -> float | None:
        if geometry.is_empty:
            return None
        nearest = nearest_points(start, geometry)[1]
        return float(start.distance(nearest))

    @staticmethod
    def _direction(yaw_deg: float) -> tuple[float, float]:
        yaw_rad = math.radians(yaw_deg)
        return math.cos(yaw_rad), math.sin(yaw_rad)

    @staticmethod
    def _clean_m(value: float) -> float:
        return 0.0 if abs(value) < 1e-12 else round(value, 12)

    @staticmethod
    def _clean_cm(value_m: float) -> float:
        value = value_m * 100.0
        return 0.0 if abs(value) < 1e-9 else round(value, 6)


def preview_world_config() -> WorldConfig:
    return WorldConfig(
        scene_id="indoor-lab-preview",
        bounds=WorldBounds(min_x=-6.0, max_x=6.0, min_y=-4.0, max_y=4.0),
        obstacles=(
            WorldObstacle(id="preview-a", center_x_m=3.0, center_y_m=-2.0, size_x_m=1.4, size_y_m=0.8),
            WorldObstacle(id="preview-b", center_x_m=-3.0, center_y_m=-2.5, size_x_m=1.2, size_y_m=1.2),
            WorldObstacle(id="preview-c", center_x_m=4.0, center_y_m=2.5, size_x_m=1.0, size_y_m=1.6),
        ),
        semantic_objects=_semantic_objects(),
    )


def obstacle_world_config(
    wide_side: Literal["left", "right", "equal"] = "left",
) -> WorldConfig:
    side_obstacles: tuple[WorldObstacle, ...]
    if wide_side == "left":
        side_obstacles = (
            WorldObstacle(id="right-crate", center_x_m=0.35, center_y_m=-1.0, size_x_m=0.8, size_y_m=0.5),
        )
    elif wide_side == "right":
        side_obstacles = (
            WorldObstacle(id="left-crate", center_x_m=0.35, center_y_m=1.0, size_x_m=0.8, size_y_m=0.5),
        )
    else:
        side_obstacles = ()
    return WorldConfig(
        scene_id=f"indoor-lab-obstacle-{wide_side}",
        bounds=WorldBounds(min_x=-6.0, max_x=6.0, min_y=-4.0, max_y=4.0),
        obstacles=(
            WorldObstacle(id="front-crate", center_x_m=1.2, center_y_m=0.0, size_x_m=0.5, size_y_m=0.8),
            *side_obstacles,
        ),
        semantic_objects=_semantic_objects(),
    )


def _semantic_objects() -> tuple[WorldSemanticObject, ...]:
    return (
        WorldSemanticObject(
            id="person-green-01",
            kind="person",
            label="green lab operator",
            color="green",
            center_x_m=-2.8,
            center_y_m=2.4,
            size_x_m=0.45,
            size_y_m=0.45,
            height_m=1.7,
            interaction_radius_m=0.9,
            blocking=True,
        ),
        WorldSemanticObject(
            id="table-gray-01",
            kind="table",
            label="gray work table",
            color="gray",
            center_x_m=-3.2,
            center_y_m=0.4,
            size_x_m=1.4,
            size_y_m=0.8,
            height_m=0.75,
            interaction_radius_m=0.9,
            blocking=True,
        ),
        WorldSemanticObject(
            id="chair-yellow-01",
            kind="chair",
            label="yellow chair",
            color="yellow",
            center_x_m=-2.5,
            center_y_m=-1.4,
            size_x_m=0.55,
            size_y_m=0.55,
            height_m=0.9,
            interaction_radius_m=0.75,
            blocking=True,
        ),
        WorldSemanticObject(
            id="door-white-01",
            kind="door",
            label="white lab door",
            color="white",
            center_x_m=4.7,
            center_y_m=-2.3,
            size_x_m=0.18,
            size_y_m=1.1,
            height_m=2.1,
            interaction_radius_m=1.0,
            blocking=True,
        ),
        WorldSemanticObject(
            id="bottle-red-01",
            kind="bottle",
            label="red bottle",
            color="red",
            center_x_m=0.0,
            center_y_m=2.4,
            size_x_m=0.12,
            size_y_m=0.12,
            height_m=0.28,
            interaction_radius_m=0.55,
            blocking=False,
        ),
        WorldSemanticObject(
            id="goal-zone-blue-01",
            kind="goal_zone",
            label="blue goal zone",
            color="blue",
            center_x_m=3.5,
            center_y_m=2.4,
            size_x_m=1.0,
            size_y_m=1.0,
            height_m=0.03,
            interaction_radius_m=0.65,
            blocking=False,
        ),
    )


__all__ = [
    "RobotSpawn",
    "WorldBounds",
    "WorldConfig",
    "WorldGeometry",
    "WorldObstacle",
    "SemanticObjectColor",
    "SemanticObjectKind",
    "WorldSemanticObject",
    "obstacle_world_config",
    "preview_world_config",
]
